sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    internal class OperationInternal<TResult>
+    {
+        private const string RetryAfterHeaderName = "Retry-After";
+
+        private readonly object SetStateLock = new object();
+
+        private readonly IOperation<TResult> _operation;
+
+        private TResult _value;
+
+        public OperationInternal(ClientDiagnostics clientDiagnostics, IOperation<TResult> operation) : this(clientDiagnostics)
+        {
+            _operation = operation;
+            OperationTypeName = operation.GetType().Name;
+        }
+
+        protected OperationInternal(ClientDiagnostics clientDiagnostics)
+        {
+            ClientDiagnostics = clientDiagnostics;
+            DefaultPollingInterval = TimeSpan.FromSeconds(1);
+            ScopeAttributes = new Dictionary<string, string>();
+        }
+
+        public IDictionary<string, string> ScopeAttributes { get; }
+
+        public bool HasValue { get; private set; }
+
+        public bool HasCompleted { get; set; }
+
+        public TResult Value
+        {
+            get
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (OperationFailedException != null)
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
```

Comment:
who would use this setter?

Reply:
Some Key Vault operations set `Value` in the operation constructor and make it available from the start:
https://github.com/Azure/azure-sdk-for-net/blob/793f822150df146c6ca624241436e82fa6ab5a6c/sdk/keyvault/Azure.Security.KeyVault.Keys/src/DeleteKeyOperation.cs#L27

`MixedReality.RemoteRendering` exposes a `Value` from start and its mutable:
https://github.com/Azure/azure-sdk-for-net/blob/793f822150df146c6ca624241436e82fa6ab5a6c/sdk/remoterendering/Azure.MixedReality.RemoteRendering/src/StartRenderingSessionOperation.cs#L45-L63

Reply:
Alternately, since in most LROs this won't happen, we could hide the setter and let these operations handle this by themselves. For example:
```cs
public TResult Value => _operationInternal.HasCompleted ? _operationInternal.Value :  _value;
```